### ISSUE TITLE
Invalid type checking error in `__getitem__`

### ISSUE THREAD

**Issue by plafosse** [microsoft/pyright#2069]
The following produces the error `No overloads for "__getitem__" match the provided arguments
  Argument types: (int | slice)`  on the last line.

```python
from typing import Union, List
class A:
    def __init__(self, l:List[int]):
       self.my_list = l
    def __getitem__(self, value:Union[str, int, slice]):
        if isinstance(value, str):
            return self.my_list[int(value)]
        if isinstance(value, (int, slice)):
            return self.my_list[value]
```

the logical equivalent of this doesn't produce an error

```python
from typing import Union, List
class A:
    def __init__(self, l:List[int]):
       self.my_list = l
    def __getitem__(self, value:Union[str, int, slice]):
        if isinstance(value, str):
            return self.my_list[int(value)]
        if isinstance(value, int):
            return self.my_list[value]
        elif isinstance(value, slice):
            return self.my_list[value]
```

**VS Code extension or command-line**
Pyright 1.1.155 via the VSCode extension

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release. Here's the [commit](https://github.com/microsoft/pyright/commit/3931ec95b74d0f812355154db110ff538e64583f) that fixes it.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.156, which I just published. It will also be included in the next release of pylance.
